src/Common/FileCache.cpp

```diff
@@ -0,0 +1,609 @@
+        auto segment_range = (*it)->range();
+
+        size_t current_pos;
+        if (segment_range.left < range.left)
+        {
+            ///    [_______     -- requested range
+            /// [_______
+            /// ^
+            /// segment1
+
+            current_pos = segment_range.right + 1;
+            ++it;
+        }
+        else
+            current_pos = range.left;
+
+        while (current_pos <= range.right && it != file_segments.end())
+        {
+            segment_range = (*it)->range();
+
+            if (current_pos == segment_range.left)
+            {
+                current_pos = segment_range.right + 1;
+                ++it;
+                continue;
+            }
+
+            assert(current_pos < segment_range.left);
+
+            auto hole_size = segment_range.left - current_pos;
+            auto * cell = addCell(key, current_pos, hole_size, FileSegment::State::EMPTY, cache_lock);
+            file_segments.insert(it, cell->file_segment);
+
+            current_pos = segment_range.right + 1;
+            ++it;
+        }
+
+        if (current_pos <= range.right)
+        {
+            ///   ________]     -- requested range
+            ///   _____]
+            ///        ^
+            /// segmentN
+
+            auto hole_size = range.right - current_pos + 1;
+            auto * cell = addCell(key, current_pos, hole_size, FileSegment::State::EMPTY, cache_lock);
+            file_segments.push_back(cell->file_segment);
+        }
+    }
+
+    assert(!file_segments.empty());
+    return FileSegmentsHolder(std::move(file_segments));
+}
+
+LRUFileCache::FileSegmentCell * LRUFileCache::addCell(
+    const Key & key, size_t offset, size_t size, FileSegment::State state, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+{
+    /// Create a file segment cell and put it in `files` map by [key][offset].
+
+    if (!size)
+        return nullptr; /// Empty files are not cached.
+
+    if (files[key].contains(offset))
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+            "Cache already exists for key: `{}`, offset: {}, size: {}", keyToStr(key), offset, size);
+
+    auto file_segment = std::make_shared<FileSegment>(offset, size, key, this, state);
+    FileSegmentCell cell(std::move(file_segment), queue);
+
+    auto & offsets = files[key];
+
+    if (offsets.empty())
+    {
+        auto key_path = path(key);
+        if (!fs::exists(key_path))
+            fs::create_directories(key_path);
+    }
+
+    auto [it, inserted] = offsets.insert({offset, std::move(cell)});
+    if (!inserted)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+            "Failed to insert into cache key: `{}`, offset: {}, size: {}", keyToStr(key), offset, size);
+
+    return &(it->second);
+}
+
+bool LRUFileCache::tryReserve(const Key & key_, size_t offset_, size_t size, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+{
+    auto removed_size = 0;
```

Comment:
Maybe check corner case when we are trying to reserve more data than the whole cache size?